### PR TITLE
Handle TyAlias in projected_ty

### LINKED ISSUE #15143: "internal error: entered unreachable code: Only adt has field"
**rust-analyzer version**:
`rust-analyzer 0.0.0 (246d11b2a 2023-06-23)` (built from current main)

**rustc version**:
`rustc 1.72.0-nightly (8084f397c 2023-06-25)`

**relevant settings**:
editor: emacs, latest lsp-mode

# The Error

I've been getting a lot of these in RA's stderr. Let me know what additional information would be helpful, and thanks!

```text
Panic context:
> publish_diagnostics

thread 'Worker' panicked at 'internal error: entered unreachable code: Only adt has field', crates/hir-ty/src/mir.rs:153:21
stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::unreachable_display
   3: hir_ty::mir::ProjectionElem<V,T>::projected_ty
   4: hir_ty::mir::borrowck::moved_out_of_ref::{{closure}}
   5: hir_ty::mir::borrowck::moved_out_of_ref
   6: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
   7: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
   8: core::iter::adapters::try_process
   9: hir_ty::mir::borrowck::borrowck_query
  10: salsa::runtime::Runtime::execute_query_implementation
  11: salsa::derived::slot::Slot<Q,MP>::read_upgrade
  12: salsa::derived::slot::Slot<Q,MP>::read
  13: <salsa::derived::DerivedStorage<Q,MP> as salsa::plumbing::QueryStorageOps<Q>>::try_fetch
  14: <DB as hir_ty::db::HirDatabase>::borrowck::__shim
  15: <DB as hir_ty::db::HirDatabase>::borrowck
  16: hir::DefWithBody::diagnostics
  17: hir::Module::diagnostics
  18: ide_diagnostics::diagnostics
  19: salsa::Cancelled::catch
  20: ide::Analysis::diagnostics
  21: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
  22: alloc::vec::in_place_collect::<impl alloc::vec::spec_from_iter::SpecFromIter<T,I> for alloc::vec::Vec<T>>::from_iter
  23: core::ops::function::FnOnce::call_once{{vtable.shim}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Panic context:
> 
version: 0.0.0 (246d11b2a 2023-06-23)
request: textDocument/codeAction CodeActionParams {
    text_document: TextDocumentIdentifier {
        uri: Url {
            scheme: "file",
            cannot_be_a_base: false,
            username: "",
            password: None,
            host: None,
            port: None,
            path: "/Users/jbr/code/divviup/divviup-api/src/entity/api_token.rs",
            query: None,
            fragment: None,
        },
    },
    range: Range {
        start: Position {
            line: 0,
            character: 0,
        },
        end: Position {
            line: 0,
            character: 0,
        },
    },
    context: CodeActionContext {
        diagnostics: [],
        only: None,
        trigger_kind: None,
    },
    work_done_progress_params: WorkDoneProgressParams {
        work_done_token: None,
    },
    partial_result_params: PartialResultParams {
        partial_result_token: None,
    },
}
```

### PR DESCRIPTION
**Pull request by alibektas**
First of all I still have no idea how MIR works but #15143 has been an issue that constantly made RA crash so I have been looking for a way to make RA stop panicking. I have zero claims that what I want to merge has any sense or is correct :smile:  but there isn't any more panicking. Even if it is wrong may this be at least a step towards resolving this issue.   
As is customary this PR fixes #15143

### CODE REVIEW

File: crates/hir-ty/src/mir.rs [hunk 1]
```diff
@@ -151,6 +151,7 @@ impl<V, T> ProjectionElem<V, T> {
                 TyKind::Adt(_, subst) => {
                     db.field_types(f.parent)[f.local_id].clone().substitute(Interner, subst)
                 }
+                TyKind::Alias(al) => al.clone().intern(Interner),
```
Comment (lnicola) [inline crates/hir-ty/src/mir.rs:154]:
```suggestion
                TyKind::Alias(inner) => al.clone().intern(Interner),
```

File: crates/hir-ty/src/mir.rs [hunk 2]
```diff
@@ -167,6 +168,7 @@ impl<V, T> ProjectionElem<V, T> {
                         TyKind::Error.intern(Interner)
                     }),
                 TyKind::Closure(id, subst) => closure_field(*id, subst, *f),
+                TyKind::Alias(al) => al.clone().intern(Interner),
```
Comment (lnicola) [inline crates/hir-ty/src/mir.rs:171]:
```suggestion
                TyKind::Alias(inner) => al.clone().intern(Interner),
```

**Comment by lnicola** [comment]
I'm not completely sure this is correct, but it can't be worse than crashing, right? :-). Can you also find a place to add a test?

**Comment by alibektas** [comment]
> I'm not completely sure this is correct, but it can't be worse than crashing, right? :-). Can you also find a place to add a test?

This was my exact line of thought :smile: . I don't seem to understand why the enclosing function is invoked and I do not see a testing infrastructure that isolates mir from the rest. I would be happy to do it if you could maybe guide me ?

**Review by lnicola — APPROVED** [review]
Fair enough, I actually don't have a minimization for this.

CC @HKalbasi maybe you know a way to test this?

**Comment by HKalbasi** [comment]
I don't think this PR makes any sense. The `projected_ty` should return the type of place after the applied projection. That is, if the base ty is `(i32, u64)` and the projection is `.1`, the result should be `u64`. This PR doesn't do this, if the goal is to silence the problem, we can just comment out the `never!` lines.

But what is the problem here? In this code (which hits the `never!` so it can be used as a unit test):
```Rust
trait Tr {
    type Ty;
}

struct A;

impl Tr for A {
    type Ty = (u32, i64);
}

struct B<T: Tr> {
    f: <T as Tr>::Ty,
}

fn x(b: B<A>) {
    let f = b.f.0; // here
}
```
`b.f.0` will become a place in mir and it tries to recursively compute its type. But the result of `b.f` is `<A as Tr>::Ty`, a `TyKind::Alias`, and it expects something with projection `.0` to have tuple type, so it fails. To solve this problem, we should either:
* Normalize the base type
* Store the type inside projection and retire the `projection_ty` function

**Comment by lnicola** [comment]
Oh, sorry, let's revert this and tell people to use a binary build, if possible.

**Pull request merged**